### PR TITLE
machines: add support for cloning VMs

### LINKED ISSUE #14523: Feature request: Clone existing virtual machines
Page: Virtual Machines

It would be nice to have a "clone" option (in the drop-down menu) for virtual machines to duplicate the settings and image. For now, it's until possible to run/shut down and delete virtual machines on this page.

image

### PR DESCRIPTION
**Pull request by KKoukiou**
Relevant to https://bugzilla.redhat.com/show_bug.cgi?id=1683391
Fixes #14523

For release notes:
Screen Shot 2020-11-24 at 12 52 13

### Machines: add support for cloning virtual machines

Cloning creates a new VM that uses its own disk image for storage, but most of the clone’s configuration and stored data is identical to the source VM. This makes it possible to prepare a number of VMs optimized for a certain task without the need to optimize each VM individually.

### CODE REVIEW

File: pkg/machines/components/vmCloneDialog.jsx [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+import { isEmpty, isObjectEmpty } from '../helpers.js';
+import { ModalError } from 'cockpit-components-inline-notification.jsx';
+
+import "./vmCloneDialog.scss";
+const _ = cockpit.gettext;
+
+export const CloneDialog = ({ name, connectionName, toggleModal }) => {
+    const [newVmName, setNewVmName] = useState(name + '-clone');
+    const [validated, setValidated] = useState(false);
```
Comment (skobyda) [inline pkg/machines/components/vmCloneDialog.jsx:32]: this state variable `validated` seems redundant. It's only changed once, to true, when `onClone` function is called (even before the variable it is used in Button `isDisabled`) and then it's never changed back to false again.

We can just remove it and nothing will change.

File: pkg/machines/components/vmCloneDialog.jsx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                .stream(setVirtCloneOutput)
+                .then(toggleModal, exc => {
+                    setInProgress(false);
+                    dialogErrorSet({ dialogError: cockpit.format(_("Failed to clone VM $0"), name), dialogErrorDetail: exc.message });
+                });
+    }
+
+    const validationFailed = validateParams();
+    return (
+        <Modal position="top" variant="medium" isOpen onClose={toggleModal}
```
Comment (skobyda) [inline pkg/machines/components/vmCloneDialog.jsx:65]: Maybe for such slender dialog with only 1 input field the varian should be small/little? Altought I'm not sure, WDYT?

File: pkg/machines/components/vmCloneDialog.jsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+           title={cockpit.format(_("Create a clone VM based on $0"), name)}
+           footer={
+               <>
+                   {!isObjectEmpty(error) && <ModalError dialogError={error.dialogError} dialogErrorDetail={error.dialogErrorDetail} />}
+                   {virtCloneOutput && <code className="vm-clone-virt-clone-output">{virtCloneOutput}</code>}
```
Comment (skobyda) [inline pkg/machines/components/vmCloneDialog.jsx:70]: The error handling isn't really useful here. When I try to cause an error by inputting the name of an existing VM, I get an error: 
```
Failed to clone VM fedora32
Error message: virt-clone exited with code 1 
```
Which isn't really useful.

When I run the same in CLI, I get an output:
`ERROR    Invalid name for new guest: Guest name 'rww' is already in use.`
Which is much more useful and that's what we want to show in a modal.

I guess that's what the `virtCloneOutput` is there for. That right now returns and empty string. Adding a `pty` parameter to cockpit.spawn should fix that ;).

**Review by skobyda — CHANGES_REQUESTED** [review]

File: pkg/machines/components/vmCloneDialog.jsx (hunk 3 above)
Comment (KKoukiou) [reply pkg/machines/components/vmCloneDialog.jsx]: Fixed.

File: pkg/machines/components/vmCloneDialog.jsx (hunk 2 above)
Comment (KKoukiou) [reply pkg/machines/components/vmCloneDialog.jsx]: Fixed.

File: pkg/machines/components/vmCloneDialog.jsx (hunk 1 above)
Comment (KKoukiou) [reply pkg/machines/components/vmCloneDialog.jsx]: Fixed.

**Review by skobyda — DISMISSED** [review]
Thanks for a nice feature

**Review by skobyda — DISMISSED** [review]

**Comment by KKoukiou** [comment]
The new test is a bit flaky  - https://logs.cockpit-project.org/logs/pull-14954-20201202-070348-51aa8601-rhel-8-3/log.html

I will repost.

**Comment by KKoukiou** [comment]
@skobyda sorry I needed to repost because the test was flaky. I just moved the test into the testBasic, because it's small feature, does not need it's own test. And I fixed the flakiness by checking for the disabled aria-label before clicking the button. Can you please reapprove?

**Review by skobyda — APPROVED** [review]

**Comment by garrett** [comment]
Why is the status in a text block instead of using a UI progress bar with details?

And cancel is still active — does that stop the cloning process? And if so, does it clean up afterward?

**Comment by KKoukiou** [comment]
> Why is the status in a text block instead of using a UI progress bar with details?
> 
> And cancel is still active — does that stop the cloning process? And if so, does it clean up afterward?
> 
> Can we address these in follow-up PRs? I'm happy to open issues if there aren't some already.

Because virt-clone is just a CLI command, and we get the stream output from it, it's not an API which provides us with a progress percentage and a text detail. So with the currently available tools this is what we can get.

I need to test the cancel, good point.

**Pull request merged**